picard/util/pipe.py

```diff
@@ -0,0 +1,241 @@
+            win32file.WriteFile(pipe, str.encode(message))
+        finally:
+            win32file.CloseHandle(pipe)
+
+        return True
+
+    def __unix_sender(self, message: str) -> bool:
+        with open(self.path, 'a') as fifo:
+            fifo.write(message)
+        return True
+
+    def send_to_pipe(self, message: str, timeout_secs: Optional[float] = None) -> bool:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        __pool = concurrent.futures.ThreadPoolExecutor()
+
+        if self.__is_win:
+            sender = __pool.submit(self.__win_sender, message)
+        else:
+            sender = __pool.submit(self.__unix_sender, message)
+
+        try:
+            if sender.result(timeout=timeout_secs):
+                return True
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the sender
+            self.read_from_pipe()
+
+        return False
+
+    def read_from_pipe(self, timeout_secs: Optional[float] = None) -> str:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        __pool = concurrent.futures.ThreadPoolExecutor()
+
+        if self.__is_win:
+            reader = __pool.submit(self.__win_reader)
+        else:
+            reader = __pool.submit(self.__unix_reader)
+
+        try:
+            if reader.result(timeout=timeout_secs):
+                res: str = reader.result()
+                if res != self.MESSAGE_TO_IGNORE:
+                    return res
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the file-opening loop
+            self.send_to_pipe(self.MESSAGE_TO_IGNORE)
+
+        return Pipe.NO_RESPONSE_MESSAGE
+
+    def __win_reader(self) -> str:
+        response = ""  # type: ignore
+
+        try:
+            pipe = win32file.CreateFile(
+                self.path,
+                win32file.GENERIC_READ | win32file.GENERIC_WRITE,
+                self.__SHARE_MODE,
+                None,
+                win32file.OPEN_EXISTING,
+                self.__FLAGS_AND_ATTRIBUTES,
+                None
+            )
+            while not response:
+                response = win32file.ReadFile(pipe, self.__BUFFER_SIZE)
+
+        except WinApiError as err:
+            if err.winerror == self.__FILE_NOT_FOUND_ERROR_CODE:
+                raise FileNotFoundError(Pipe.NOT_FOUND_MESSAGE)
+            elif err.winerror == self.__BROKEN_PIPE_ERROR_CODE:
+                raise FileNotFoundError("Pipe is broken")
+            else:
+                raise FileNotFoundError(f"{err.winerror}; {err.funcname}; {err.strerror}")
+
+        # response[0] stores an exit code while response[1] an actual response
+        if response:
+            if response[0] == 0:
+                return str(response[1].decode("utf-8"))  # type: ignore
+            else:
+                raise ValueError(f"INVALID RESPONSE: {response[1].decode('utf-8')}")  # type: ignore
```

Comment:
What about using our own exception here?

```python
class PipeError(Exception):
    pass

class PipeErrorInvalidResponse(PipeError):
   pass


...

    raise PipeErrorInvalidResponse(f"INVALID RESPONSE: {response[1].decode('utf-8')}")
```

It could be extended, and instead of raising FileNotFoundError everywhere we could raise specific exceptions.

Reply:
Making our own errors and storing error messages there is a great idea, thanks!

Reply:
I've created our custom errors but I'm open to any ideas to avoid the boilerplate there.